### PR TITLE
start on micro quark

### PR DESCRIPTION
**Pull request by ztatlock**
Let's try to build "micro Quark" in this branch.

It should be a super stripped down version. It doesn't even need to use WebKit.

The main thing we want to figure out is how the trace spec looks and how we can prove things about it.

@Ptival can you take ownership of this? Let's meet this morning to get started on it.

### CODE REVIEW

**Comment by ztatlock** [comment]
@Ptival What if we make the component type correspond to the component's executable? e.g.

``` c

Init {
  spawn("UserInput");
  spawn("AddressBar");
}

Exchange {
  UserInput {
     ... handlers ...
  }
  AddressBar {
     ... handlers ...
  }
}
```

**Comment by Ptival** [comment]
Is it the case that components are always created through an executable though? (like, for instance, how are tabs created?)

**Comment by ztatlock** [comment]
Well, we can do things however we like.

But yes, right now I think all components will be created by spawn.

For example, tab creation could be like:

``` c
Messages {
  AddTab(str);
}

Init {
  spawn("UserInput");
  spawn("AddressBar");
}

Exchange {
  UserInput {
    AddTab(url) {
      spawn("Tab");
      // TODO we need some way to associate url with this tab
    }
     ... handlers ...
  }
  AddressBar {
     ... handlers ...
  }
  Tab {
    ... handlers ...
  }
}
```

Perhaps we should add a "Components" section to the DSL where the user defines all the paths to component executables and gives them names? e.g.

``` c
Components {
  UserInput := "user-input.py";
  Tab = "webkit-tab.py";
  AddressBar = "addr-bar";
}
```

Definitely feels crufty, but it may ease code generation for now...

**Comment by Ptival** [comment]
I have a small problem figuring out the extracted code.

The thing is, I would have wished to define the state as a record:

``` ocaml
Record kstate :=
{ (* here the usual *)
; userinput : list chan * handler
; tabs : list chan * handler
}.
```

But with handler defined as:

``` ocaml
Definition handler := forall (c : chan) (s : kstate),
  STsep (kstate_inv kst * [In c (components kst)]) (fun (kst' : kstate) => kstate_inv kst').
```

But this creates all sort of inter-dependencies, since it refers both to kstate and kstate_inv.

Any thought?

**Comment by ztatlock** [comment]
What happens if we just move all the Trace predicates (kstate_inv included) to earlier in the extracted code? Will that work?

**Comment by Ptival** [comment]
No, the problem is not quite the order of things as much as how they are defined.

We can't have the state holding handlers whose type mentions an invariant over such states.

Either we loosen up the definition of handlers, so that they are not about holding an invariant over states, or we keep things separate somehow...

**Comment by ztatlock** [comment]
One solution: don't put handlers in the kstate.

Instead make a component type type:

``` ocaml
Record kstate :=
{ (* here the usual *)
; components : list (component_t * list chan)
; ...
}.
```

Then have a dispatcher call the right handler based on the component type.

That way, handler can have the type you want and we break the dependency.

Does that sound reasonable?

**Comment by Ptival** [comment]
That sounds reasonable.

**Comment by ztatlock** [comment]
@Ptival Should I look into that bug you mentioned from the stateful branch?

 What exactly wasn't working?

**Comment by Ptival** [comment]
So, it turns out that this is _the_ pull request from u-quark into master, and no other pull request from the same branch can exist at one time.

Let's try and create a pull request from stateful to master and see how this reacts! :)

**Comment by ztatlock** [comment]
You've been making great progress on this!

**Comment by Ptival** [comment]
``` bash
State {
  currenttab : fdesc ;
  renderer   : fdesc ;
}

Components {
  Renderer  "renderer.py"  ;
  Tab       "tab.py"       ;
  UserInput "userinput.py" ;
}

Messages {
  Dummy(str);
  Input(str);
  Tab(str);
  Go(str);
  HTML(str);
}

Init {
  spawn(Renderer);
  spawn(Tab);
  spawn(UserInput);
}

Exchange(c) {
  Renderer {
    # A dummy message is sent on startup so that the chan can be stored
    Dummy(s) {
      renderer := c;
    }
  }

  Tab {
    # A dummy message is sent on startup so that the chan can be stored
    Dummy(s) {
      currenttab := c;
    }

    HTML(s) {
      # TODO: Check that it's the current tab
      send(renderer, HTML(s));
    }
  }

  UserInput {
    Input(s) {
      send(currenttab, Go(s));
    }
  }
}
```

What think?

**Comment by ztatlock** [comment]
```
State {
  screen : fdesc;
  cur_tab : fdesc;
}

Components {
  Tab       "tab.py";
  Screen    "screen.py";
  UserInput "user-input.py";
}

Messages {
  Hello();
  Input(str);
  Wget(str);
  HTML(str, fdesc);
  Display(fdesc); 
}

Init {
  spawn(Tab);
  spawn(Screen);
  spawn(UserInput);
}

Exchange(c) {
  Tab {
    Hello() {
      cur_tab := c;
    }
    Wget(url) {
      f = call("wget.py");
      send(c, HTML(url, f));
    }
    Display(bitmap) {
      # TODO check that this is from cur tab
      send(screen, Display(bitmap)); 
    } 
  }

  Screen {
    Hello() {
      screen := c;
    }
  } 

  UserInput {
    Input(i) {
      send(cur_tab, Input(i));
    }
  }
} 
```

How about this?

Do the differences make sense?

**Comment by ztatlock** [comment]
ooops...

**Comment by Ptival** [comment]
Hmm, I need to add a Chan type.

@ztatlock Could you describe the changes needed in genC.ml and the subsequent changes needed on the C binding side?

**Comment by Ptival** [comment]
@ztatlock Just merged master into u-quark, and it solved my passfd problem, great!

@d1jang Commit e8e0bc3 fixes your issue, but commit ad28f83 is in a cleaner state. I suggest you rebase your u-ssh branch onto the latter (which, at the moment, is where the u-quark branch points to). If you feel incomfortable about whether and how to do this, we can go over it tomorrow.

It is now time to make u-quark work!

File: compiler/src/genCoq.ml [hunk 1]
```diff
@@ -17,16 +17,19 @@ let coq_of_typ = function
   | Num -> "num"
   | Str -> "str"
   | Fdesc -> "fdesc"
+  | Chan -> "tchan"
```
Comment (ztatlock) [inline compiler/src/genCoq.ml:20]: Ideally, adding chans could have been a separate PR, but no big deal.

File: compiler/src/genCoq.ml [hunk 2]
```diff
@@ -281,6 +284,29 @@ let coq_of_msg_pat m =
 let handler_vars xch_chan h =
   uniq (xch_chan :: h.trigger.payload @ prog_vars h.respond)
 
+module OrderedChan =
+  struct
+    type t = chan
+    let compare = compare
+  end
+
+module ChanSet = Set.Make(OrderedChan);;
```
Comment (ztatlock) [inline compiler/src/genCoq.ml:293]: nice

File: compiler/src/genCoq.ml [hunk 3]
```diff
@@ -324,12 +358,24 @@ let coq_of_handler s xch_chan h =
...
-    [ mkstr "[In %s comps]" xch_chan
-    ; mkstr "bound %s" xch_chan
-    ; mkstr "all_bound_drop comps %s" xch_chan
-    ; "(tr ~~ [KTrace tr])"
-    ]
+  let fr = (
+    let chans = chans_of_prog h.respond in
+    frames_of_vars s @ (
+      List.fold_left
+      (* folding function *)
```
Comment (ztatlock) [inline compiler/src/genCoq.ml]: Why not

``` ocaml
(fun acc elt -> mkstr "bound %s" elt :: acc)
```

(again, just a nitpick, not a big deal -- it makes it 1 line, then you prolly don't need comment? maybe...)

**Comment by ztatlock** [comment]
Looks pretty good to me.

Is the next step to merge master into this branch so that the merge can be accomplished automatically?

File: compiler/src/genCoq.ml (hunk 3 above)
Comment (Ptival) [inline compiler/src/genCoq.ml]: Yep right. At some point it was: 

``` OCaml
mkstr "bound %s" elt
:: mkstr "[In %s comps]" elt
:: acc

or something even longer maybe, so I made lines, but now that it's simple again, your point is valid.
```

**Comment by Ptival** [comment]
Ready to merge into master @ztatlock ?

**Comment by ztatlock** [comment]
Merge at will @Ptival !

**Pull request merged**